Tree Tab 2.0

ghickman opened the pull request:
This rebuilds the Codelist version page's Tree tab with a modified version of the CodelistBuilder component.  It turned out that only `TermAndCode` was needed from CodelistBuilder so I've left the rest of the sub components where they are.

This also pulls a couple of `tree_tables`-related structures into the coding system specific modules (so we can support CTV3 and SNOMED).  Most notbably, concept codes grouped by type are now generated in the coding system (since CTV3 codes can have more than one group) and passed into `tree_tables`.

These changes brought up the single code Hierarchy edge case (#193), which in turn brought up the three view tests (checking redirects work as expected) that broke due to this work.  I've reduced their scope to only check they are redirecting to the correct location so they're less brittle in the future.

**Aplastic Anaemia (SNOMED)**
![](https://p198.p4.n0.cdn.getcloudapp.com/items/Qwu0mrQQ/Image%202020-09-16%20at%203.21.06%20pm.png?v=859f753c81dbb03eb3a1295a0a079529)

**Aplastic Anaemia**
![](https://p198.p4.n0.cdn.getcloudapp.com/items/o0u8w0ge/Image%202020-09-16%20at%203.22.23%20pm.png?v=d592f7dee6a9492b034c8ec072cd1966)

Fixes #100 
Fixes #165

inglesp reviewed (changes requested):
Lots of good stuff here, but the tree view is no longer showing whether a code is in the codelist or not!

ghickman commented on coding_systems/ctv3/coding_system.py:
```diff
@@ -63,3 +63,32 @@ def descendant_relationships(codes):
+        type_nodes = hierarchy.ancestors(code) & types
+
+        # TODO: can we pass in codes bucketed by type to tree_tables?
+        first = list(type_nodes)[0]
```
Quite possibly (sets are unordered IIRC?), but this went away in a future commit 🙂

ghickman commented on static/src/js/tree/index.jsx:
```diff
@@ -0,0 +1,20 @@
+
+import ReactDOM from "react-dom";
+// Although React is not used in this module, if this line is removed, there is
+// a runtime error in codelistbuilder.jsx.
```
In the `ReactDOM.render()` call we pass a JSX element (`<Tree>` here), that's syntactic sugar for `React.createElement()` so we have to have the `React` import.

ghickman commented:
Black, blue, and underlined codes reinstated.  I've also fixed a bug with alignment of the descendant toggles and pipes.

I removed the `TermAndCode` component in favour of separate components for each of its constituent parts.  Both the builder and tree want to display the items like this:

```
[ N pipes ] [ +/- /? ] [ Term ] [ Code ]
```

where the `+/-/?` section is Builder specific and the `Term` needs to be optionally linked.  So by breaking out these parts both the `CodelistBuilder` and `Tree` components can compose them as necesary.

### Codelist Builder
![](https://p198.p4.n0.cdn.getcloudapp.com/items/z8uZzdNL/Image%202020-09-18%20at%2010.57.02%20am.png?v=adb6f91282bce94326ce421c8a97c593)

### Tree (CTV3)
![](https://p198.p4.n0.cdn.getcloudapp.com/items/WnuJOXzD/Image%202020-09-18%20at%2011.01.22%20am.png?v=c9fe8fbe0bf39557959604d9cda12c98)

### Tree (SNOMED)
![](https://p198.p4.n0.cdn.getcloudapp.com/items/NQur6BER/Image%202020-09-18%20at%2011.01.49%20am.png?v=951bc76ebf99b3e26cd2752069095e36)

inglesp commented on templates/codelists/version.html:
```diff
@@ -344,6 +344,8 @@ <h5 class="modal-title" id="publish-version-label">Publish Version</h5>
 
 {{ child_map|json_script:"child-map" }}
+{{ code_to_url|json_script:"codeToURL" }}
+{{ codes_in_definition|json_script:"definitionCodes" }}
```
Can this be `codesInDefinition`?

ghickman replied:
The JS variable?

inglesp replied:
Yeah -- that would make it match the pattern in all the other `json_script`s.

ghickman replied:
Fixed in 59f3365

inglesp commented on static/src/js/tree/Tree.jsx:
```diff
@@ -30,6 +30,28 @@ class Tree extends React.Component {
+
+    // conditionally wrap Term in an <href> element if there's a URL for it. We
+    // only want to link to codes included in the codelist so some codes will
+    // stay unlinked.
```
> We only want to link to codes included in the codelist so some codes will stay unlinked.

Why do we want this?

Looks like this might be a result of building up `code_to_url` in the view.  I don't think we need to do this, since the URL can be determined from the coding system ID (which the JS doesn't yet have) and the concept code (which it does have).  As such, I think it's worth hard coding the URL structure in the JS.

ghickman replied:
> Why do we want this?

I might have misunderstood something here, I'd assumed we only wanted the included codes linked.  Should I have linked all the codes?

inglesp replied:
That matches the existing behaviour.

ghickman replied:
Fixed in 67abb8a

inglesp reviewed (approved):
Looks good.  A minor tweak suggested.
